udata_piwik/metrics.py

```diff
@@ -1,123 +1,170 @@
-    oid = obj.id if hasattr(obj, 'id') else obj
-    if not isinstance(day, str):
-        day = (day or date.today()).isoformat()
-    commands = dict(('inc__values__{0}'.format(k), data[k]) for k in KEYS)
-    metrics = Metrics.objects(object_id=oid, level='daily', date=day)
-    return metrics.update_one(upsert=True, **commands)
-
-
-def clear_metrics_for_day(day):
-    if not isinstance(day, str):
-        day = (day or date.today()).isoformat()
-    commands = dict(('unset__values__{0}'.format(k), 1) for k in KEYS)
-    metrics = Metrics.objects(level='daily', date=day)
-    return metrics.update(upsert=False, **commands)
+def update_metrics_from_backend():
+    '''
+    TODO: factorize those fns (maybe class based?)
+    '''
+    update_resources_metrics_from_backend()
+    update_datasets_metrics_from_backend()
+
+
+def update_resources_metrics_from_backend():
+    '''
+    Update resources' metrics from backend.
+
+    Get a sum of all metrics for `resource_views` on the backend and
+    attach them to `resource.metrics`.
+    '''
+    log.info('Updating resources metrics from backend...')
+    client = metrics_client_factory()
+    result = client.get_views_from_all_datasets()
+    for (_, keys), _values in result.items():
+        values = next(_values)
+        values.pop('time')
+        resource_id = keys['resource']
+        try:
+            resource_id = uuid.UUID(resource_id)
+            resource = get_resource(resource_id)
+        except Exception as e:
+            log.exception(e)
+            continue
+        if resource:
+            log.debug('Found resource %s: %s', resource.id, values)
+            resource.metrics.update(values)
+            try:
+                # TODO: disable signals
+                resource.save()
+            except Exception as e:
+                log.exception(e)
+                continue
+        else:
+            log.error('Resource not found - id %s', resource_id)
+
+
+def update_datasets_metrics_from_backend():
+    '''
+    Update datasets' metrics from backend.
+
+    Get a sum of all metrics for `dataset_views` on the backend and
+    attach them to `dataset.metrics`.
+    '''
+    log.info('Updating datasets metrics from backend...')
+    client = metrics_client_factory()
+    result = client.get_views_from_all_datasets()
+    for (_, keys), _values in result.items():
+        values = next(_values)
+        values.pop('time')
+        dataset_id = keys['dataset']
+        dataset = Dataset.objects.filter(id=dataset_id).first()
+        if dataset:
+            log.debug('Found dataset %s: %s', dataset.id, values)
+            dataset.metrics.update(values)
+            try:
+                # TODO: disable signals
+                dataset.save()
+            except Exception as e:
+                log.exception(e)
+                continue
+        else:
+            log.error('Dataset not found - id %s', dataset_id)
+
+
+def upsert_metric_for_resource(resource, dataset, day, data):
```

Comment:
@quaxsze maybe move the `upsert_metric_*` and `update_*_metrics_from_backend` to different files for more clarity.